sulong/projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/intrinsics/multithreading/LLVMPThreadIntrinsics.java

```diff
@@ -0,0 +1,847 @@
+            return 0;
+        }
+    }
+
+    @NodeChild(type = LLVMExpressionNode.class)
+    @NodeChild(type = LLVMExpressionNode.class)
+    public abstract static class LLVMPThreadMutexInit extends LLVMBuiltin {
+        @Child
+        LLVMLoadNode read = null;
+
+        @Specialization
+        protected int doIntrinsic(VirtualFrame frame, Object mutex, Object attr) {
+            if (read == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                read = LLVMLanguage.getLLVMContextReference().get().getNodeFactory().createLoadNode(LLVMInteropType.ValueKind.I32);
+            }
+            // we can use the address of the native pointer here, bc a mutex
+            // must only work when using the original variable, not a copy
+            // so the address may never change
+            long mutexAddress = ((LLVMNativePointer) mutex).asNative();
+            Object mutObj = LLVMLanguage.getLLVMContextReference().get().mutexStorage.get(mutexAddress);
+            int attrValue = 0;
+
+            if (!((LLVMPointer) attr).isNull()) {
+                attrValue = (int) read.executeWithTarget(attr);
+            }
+            Mutex.Type mutexType = Mutex.Type.DEFAULT_NORMAL;
+            if (attrValue == 1) {
+                mutexType = Mutex.Type.RECURSIVE;
+            } else if (attrValue == 2) {
+                mutexType = Mutex.Type.ERRORCHECK;
+            }
+            if (mutObj == null) {
+                LLVMLanguage.getLLVMContextReference().get().mutexStorage.put(mutexAddress, new Mutex(mutexType));
+            }
+            return 0;
+        }
+    }
+
+    @NodeChild(type = LLVMExpressionNode.class)
+    public abstract static class LLVMPThreadMutexLock extends LLVMBuiltin {
+        @Specialization
+        protected int doIntrinsic(VirtualFrame frame, Object mutex) {
+            long mutexAddress = ((LLVMNativePointer) mutex).asNative();
+            Mutex mutexObj = (Mutex) LLVMLanguage.getLLVMContextReference().get().mutexStorage.get(mutexAddress);
+            if (mutexObj == null) {
+                // mutex is not initialized
+                // but it works anyway on most implementations
+                // so we will make it work here too, just using default type
+                // set the internLock counter to 1
+                mutexObj = new Mutex(Mutex.Type.DEFAULT_NORMAL);
+                LLVMLanguage.getLLVMContextReference().get().mutexStorage.put(mutexAddress, mutexObj);
+            }
+            // TODO: error code handling
+            return mutexObj.lock() ? 0 : 15;
+        }
+    }
+
+    @NodeChild(type = LLVMExpressionNode.class)
+    public abstract static class LLVMPThreadMutexTrylock extends LLVMBuiltin {
+        @Specialization
+        protected int doIntrinsic(VirtualFrame frame, Object mutex) {
+            long mutexAddress = ((LLVMNativePointer) mutex).asNative();
```

Comment:
Casting something to `LLVMNativePointer` must always be done using `LLVMNativePointer#cast`, but if you expect `mutex` to be one then you can just use that as type in the function header.